File: src/_4_tree/java/Element.java
package _4_tree.java;

public abstract class Element<Key extends Comparable<Key>, Value> {
    private final Key key;

    private final Value value;

    public Element(Key key, Value value) {
        this.key = key;
        this.value = value;
    }

    public Key getKey() {
        return key;
    }

    public Value getValue() {
        return value;
    }

    public Value search(Key key) {
        if (key != null ? key.equals(getKey()) : getKey() == null) return getValue();
        if (this instanceof Leaf) {
            return null;
        }
        if (this instanceof Tree) {
            Tree<Key, Value> tree = (Tree<Key, Value>) this;
            if (tree.getLeft() != null) {
                Value fromLeft = tree.getLeft().search(key);
                if (fromLeft != null) return fromLeft;
            }
            if (tree.getRight() != null) {
                return tree.getRight().search(key);
            }
        }
        return null;
    }

    public Value binarySearch(Key key) {
        if (key != null ? key.equals(getKey()) : getKey() == null) return getValue();
        if (this instanceof Leaf) {
            return null;
        }
        if (this instanceof Tree) {
            Tree<Key, Value> tree = (Tree<Key, Value>) this;
            if (key != null && key.compareTo(getKey()) < 0 && tree.getLeft() != null) {
                return tree.getLeft().search(key);
            }
            else if (key != null && key.compareTo(getKey()) > 0 && tree.getRight() != null) {
                return tree.getRight().search(key);
            }
        }
        return null;
    }
}

File: src/_4_tree/kotlin/Element.kt
package _4_tree.kotlin

// Algebraic type, generic, use-site variance (out)
sealed class Element<Key, Value>(val key: Key, val value: Value) {
    class Leaf<Key, Value>(key: Key, value: Value) : Element<Key, Value>(key, value)

    class Tree<Key, Value>(
            key: Key, value: Value,
            val left: Element<Key, Value>,
            val right: Element<Key, Value>? = null
    ) : Element<Key, Value>(key, value)
}

fun <K, V> Element<K, V>.search(key: K): V? =
        if (this.key == key) value
        // Exhaustive when
        else when (this) {
            is Element.Leaf -> null
            is Element.Tree -> this.left.search(key) ?: right?.search(key)
        }

fun <K : Comparable<K>, V> Element<K, V>.binSearch(key: K): V? =
        if (this.key == key) value
        else when (this) {
            is Element.Leaf -> null
            // operator overload (<)
            is Element.Tree -> if (key < this.key) left.binSearch(key) else right?.binSearch(key)
        }
